### PR TITLE
Core/Utils: Create std::optional helper class

### PR DESCRIPTION
**Pull request by FALL1N1**
quite obvious

### CODE REVIEW

File: src/common/Utilities/Optional.h [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-GPL2
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
+ * Copyright (C) 2005-2009 MaNGOS <http://getmangos.com/>
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Optional.h:4]: is this taken from tc or mangos?

**Comment by Winfidonarleyan** [comment]
Why do need it?

File: src/common/Utilities/Optional.h (hunk 1 above)
Comment (FALL1N1) [reply src/common/Utilities/Optional.h:4]: I just copy/paste the first AC copyright header I see everywhere, it's from TC tho.

Shall I remove the MaNGOS one or leave it as-is?

**Comment by FALL1N1** [comment]
> 
> 
> Why do need it?

Most importantly - cleaner code. https://en.cppreference.com/w/cpp/utility/optional

Example:
void plsDoNotBreak(Player* player, Creature* creature, uint32 something = 0, int8 somethingelse = 0)

will become something like

void plsDoNotBreak(Player* player, Creature* creature, uint32 Optional<something>, int8 Optional<somethingelse>)



I need it for upcomming commits:

TargetedMovementGenerator split to Follow/Chase movement gens.
Pet movement rewrite
GUID rewrite

File: src/common/Utilities/Optional.h (hunk 1 above)
Comment (FrancescoBorzi) [reply src/common/Utilities/Optional.h:4]: if it's from TC (and not from MaNGOS) you can remove the MaNGOS line.

Just FYI: for new files in AC (that are not imported from TC/cmangos) we have another header

File: src/common/Utilities/Optional.h [hunk 2]
```diff
@@ -0,0 +1,15 @@
+/*
+ * Copyright (C) 2016+     AzerothCore <www.azerothcore.org>, released under GNU GPL v2 license: https://github.com/azerothcore/azerothcore-wotlk/blob/master/LICENSE-GPL2
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/> 
+ */
+
+#ifndef acore_optional_h__
+#define acore_optional_h__
```
Comment (FrancescoBorzi) [inline src/common/Utilities/Optional.h:7]: for consistency with existing files, shouldn't this be uppercase?

**Review by FrancescoBorzi — APPROVED** [review]

**Comment by FrancescoBorzi** [comment]
@FALL1N1 let us know if you need this to be merged faster

**Comment by FALL1N1** [comment]
@FrancescoBorzi thank you, it's whatever as I'm not ready with the rest PRs.

**Comment by Winfidonarleyan** [comment]
You don't understand me, but I have no desire to explain :/

**Comment by Viste** [comment]
> > Why do need it?
> 
> Most importantly - cleaner code. https://en.cppreference.com/w/cpp/utility/optional, benefits from empty optimization (objects are only constructed if needed).
> 
> Example:
> `void plsDoNotBreak(Player* player, Creature* creature, uint32 something = 0, int8 somethingelse = 0)`
> 
> will become something like
> 
> `void plsDoNotBreak(Player* player, Creature* creature, Optional<uint32> something, Optional<int8> somethingelse)`
> 
> I need it for upcomming commits:
> 
> TargetedMovementGenerator split to Follow/Chase movement gens.
> Pet movement rewrite
> GUID rewrite
> Clustering

```
void plsDoNotBreak(Player* player, Creature* creature, std::optional<uint32> something, std::optional<int8> somethingelse)```
?

**Comment by FALL1N1** [comment]
> 
> 
> > > Why do need it?
> > 
> > 
> > Most importantly - cleaner code. https://en.cppreference.com/w/cpp/utility/optional, benefits from empty optimization (objects are only constructed if needed).
> > Example:
> > `void plsDoNotBreak(Player* player, Creature* creature, uint32 something = 0, int8 somethingelse = 0)`
> > will become something like
> > `void plsDoNotBreak(Player* player, Creature* creature, Optional<uint32> something, Optional<int8> somethingelse)`
> > I need it for upcomming commits:
> > TargetedMovementGenerator split to Follow/Chase movement gens.
> > Pet movement rewrite
> > GUID rewrite
> > Clustering
> 
> ```
> void plsDoNotBreak(Player* player, Creature* creature, std::optional<uint32> something, std::optional<int8> somethingelse)
> ```
> 
> ?

That's the main point, yes, however with the class it'd be noob-friendly as well as easier to merge/diff with other source codes out there (as it's used in Trinity).

Personally I don't care how it looks like, I'm doing this PR so it's easier to merge scripts in the future from TrinityCore with less code-wise differences, but yeah, it's completely optional so everyone can vote.

**Comment by Viste** [comment]
**as it's used in Trinity**
from this you  must start and(and finish i guess) say it like " i need like in TC for ctrl+c ctrl+v commits"

**Comment by FALL1N1** [comment]
> 
> 
> **as it's used in Trinity**
> from this you must start and(and finish i guess) say it like `i need like in TC for ctrl+c ctrl+v commits`

Pretty much, yeah :D

The main reason for me personally is that I do not want to waste time reading the c++ functions (as obviously I do not know/remember them all), time that I do not want to invest in PRs due to leeches /they could at least test and give some feedback/.

**Review by Si1ker — APPROVED** [review]

**Pull request merged**